File: ol-swift/ol-swift/olsource.swift
//
//  ol.swift
//  ol-swift
//
//  Created by 伍 威 on 14/12/29.
//  Copyright (c) 2014年 sleepsaint. All rights reserved.
//

import Foundation

extension OL {
    
    class Source {
        let source: String
        var cursor: String.Index
        var token: Character = "\0"
        var errorLog = ""
        var tokenString: String = ""
        var tokenNumber: Double = 0
        init(source: String) {
            self.source = source
            self.cursor = source.startIndex
            nextToken()
        }
        func nextToken() {
            var start = cursor;
            if cursor < source.endIndex {
                let c = source[cursor]
                switch c {
                case "~", "!", "@", "^", "(", ")", "{", "}", ",", "#", ".":
                    token = c;
                    ++cursor;
                case " ", "\n":
                    ++cursor
                    nextToken()
                case "$":
                    token = "n"
                    ++cursor;
                    start = cursor;
                    PP: while cursor < source.endIndex {
                        switch source[cursor] {
                        case "~", "!", "@", "^", "(", ")", "{", "}", ",", "#", " ", "\n", "$":
                            break PP;
                        default:
                            ++cursor;
                        }
                    }
                    tokenNumber = (source.substringWithRange(start ..< cursor) as NSString).doubleValue;
                default:
                    token = "s"
                    unescape()
                }
            } else {
                token = "\0"
            }
        }
        func unescape() {
            tokenString.removeAll(keepCapacity: true);
            while cursor < source.endIndex {
                let c = source[cursor]
                switch c {
                case "~", "!", "@", "^", "(", ")", "{", "}", ",", "#", ".", "$":
                    return
                case "\\":
                    ++cursor
                    let d = source[cursor]
                    switch d {
                    case "\\":
                        tokenString.append(Character("\\"))
                    case "n":
                        tokenString.append(Character("\n"))
                    case "r":
                        tokenString.append(Character("\r"))
                    case "t":
                        tokenString.append(Character("\t"))
                    default:
                        tokenString.append(d)
                    }
                    ++cursor
                default:
                    tokenString.append(c)
                    ++cursor;
                }
            }
        }
        func match(expected: Character) -> Bool {
            if expected == token {
                nextToken()
                return true
            } else {
                return false
            }
        }
        func error(e: String) {
            if errorLog.isEmpty {
                errorLog = "\(e) at \(cursor)"
            }
        }
        func getNumber() -> Value? {
            if match("n") {
                return Number(value: tokenNumber)
            }
            return nil
        }
        func getString() -> Value? {
            if match("s") {
                return String2(value: tokenString)
            }
            return nil
        }
        func getPath() -> Value? {
            let root = token
            if match("^") || match("~") || match("@") {
                let ret = Path(root: root)
                while match(".") {
                    if let key = getKey() {
                        ret.addKey(key)
                    } else {
                        error("can not match a key")
                        return nil
                    }
                }
                return ret
            }
            return nil
        }
        
        func getKey() -> Value? {
            return getString() ?? getFragment() ?? getList()
        }
        func getFragment() -> Value? {
            if match("{") {
                if let ret = getPath() {
                    if match("}") {
                        return ret
                    } else {
                        error("can not match }")
                    }
                } else {
                    error("in {} must a path")
                }
            }
            return nil
        }
        func getList() -> Value? {
            if match("(") {
                if let head = getValue() {
                    let ret = List(head: head)
                    while match(",") {
                        if let item = getValue() {
                            ret.addItem(item)
                        } else {
                            error("tail can not match a value")
                            return nil
                        }
                    }
                    if match(")") {
                        return ret
                    } else {
                        error("can not match )")
                    }
                }
            }
            return nil
        }
        func getNegative() -> Value? {
            if match("!") {
                if let value = getValue() {
                    return Negative(value: value)
                } else {
                    error("can not match value for !")
                }
            }
            return nil
        }
        func getOuote() -> Value? {
            if match("#") {
                if let value = getValue() {
                    return Quote(value: value)
                } else {
                    error("can not match value for #")
                }
            }
            return nil
        }
        func getValue() -> Value? {
            return getString() ?? getNumber() ?? getPath() ?? getList() ?? getNegative() ?? getOuote()
        }
    }
    
    static func parse(source: String) -> Value? {
        let s = Source(source: source)
        if let ret = s.getValue() {
            return ret
        } else {
            println("parse error: \(source)")
            println(s.errorLog)
            return nil
        }
    }
}

File: ol-cpp/ol-cpp/olsource.cpp
//
//  olsource.cpp
//  ol-cpp
//
//  Created by 伍威 on 14/12/31.
//  Copyright (c) 2014年 sleepsaint. All rights reserved.
//

#include <iostream>
#include <sstream>
#include "olsource.h"

using namespace std;

namespace OL {
    
    enum ScanStatus {
        None, String, Number
    };
    
    bool Source::debug = false;
    
    Source::Source(const char* source, size_t length)
    : _source(source), _end(source + length), _cursor(source), _token(0) {
        nextToken();
    }
    
    void Source::nextToken() {
        if (_cursor < _end) {
            auto c = *_cursor;
            const char* start;
            switch (c) {
                case '^':
                case '@':
                case '~':
                case '!':
                case '{':
                case '}':
                case '(':
                case ')':
                case ',':
                case '#':
                case '.':
                    _token = c;
                    ++_cursor;
                    break;
                case '\n':
                case ' ':
                    ++_cursor;
                    nextToken();
                    break;
                case '$':
                    _token = NUMBER_TOKEN;
                    ++_cursor;
                    start = _cursor;
                    while (_cursor < _end) {
                        switch (*_cursor) {
                            case '^':
                            case '@':
                            case '~':
                            case '!':
                            case '{':
                            case '}':
                            case '(':
                            case ')':
                            case ',':
                            case '#':
                            case '\n':
                            case ' ':
                            case '$':
                                goto PP;
                            default:
                                ++_cursor;
                        }
                    }
                PP: _tokenNumber = atof(start);
                    break;
                default:
                    _token = STRING_TOKEN;
                    unescape();
            }
        } else {
            _token = 0;
        }
    }
    void Source::unescape() {
        _tokenString.reset();
        while (_cursor < _end) {
            auto c = *_cursor;
            char d;
            switch (c) {
                case '^':
                case '@':
                case '~':
                case '!':
                case '{':
                case '}':
                case '(':
                case ')':
                case ',':
                case '#':
                case '.':
                case '$':
                    return;
                case '\\':
                    ++_cursor;
                    d = *_cursor;
                    switch (d) {
                        case '\\':
                            _tokenString.append('\\');
                            break;
                        case 'n':
                            _tokenString.append('\n');
                            break;
                        case 'r':
                            _tokenString.append('\r');
                            break;
                        case 't':
                            _tokenString.append('\t');
                            break;
                        default:
                            _tokenString.append(d);
                    }
                    ++_cursor;
                    break;
                default:
                    _tokenString.append(c);
                    ++_cursor;
                    break;
            }
        }
    }
    
    void Source::error(const std::string &e) {
        if (debug && _errorLog.empty()) {
            ostringstream s;
            s <<  e <<  " at " << (_cursor - _source);
            _errorLog = s.str();
        }
    }
    
    Quote* Source::getNumber() {
        if (match(NUMBER_TOKEN)) {
            return new Quote(new class Number(_tokenNumber));
        }
        return nullptr;
    }
    
    Quote* Source::getString() {
        if (match(STRING_TOKEN)) {
            return new Quote(new class String(_tokenString.begin(), _tokenString.end()));
        }
        return nullptr;
    }
    
    Path* Source::getPath(int root) {
        Path* path = new Path(root);
        while (match('.')) {
            Value* key = getKey();
            if (key) {
                path->append(key);
            } else {
                error("can not match a key");
                path->release();
                return nullptr;
            }
        }
        return path;
    }
    
    Path* Source::getFragment() {
        auto token = _token;
        Path* path;
        nextToken();
        switch (token) {
            case '^':
            case '@':
            case '~':
                path = getPath(token);
                if (path) {
                    if (match('}')) {
                        return path;
                    } else {
                        error("can not match }");
                        path->release();
                    }
                } else {
                    error("in {} must a path");
                }
            default:
                break;
        }
        return nullptr;
    }
    
    List* Source::getList() {
        Value* head = getValue();
        if (head) {
            List* list = new List(head);
            while (match(',')) {
                Value* item = getValue();
                if (item) {
                    list->append(item);
                } else {
                    error("tail can not match a value");
                    break;
                }
            }
            if (match(')')) {
                return list;
            } else {
                error("can not match )");
            }
            list->release();
        }
        return nullptr;
    }
    
    Value* Source::parse(const char *source, size_t length) {
        Source s = Source(source, length);
        auto ret = s.getValue();
        if (debug && !ret) {
            cout << string(source, length) << endl;
            cout << s._errorLog << endl;
        }
        return ret->autoRelease();
    }
    
    Value* Source::parse(const std::string &source) {
        return parse(source.c_str(), source.length());
    }
    
    Value* Source::getValue() {
        auto token = _token;
        Value* value;
        nextToken();
        switch (token) {
            case STRING_TOKEN:
                return new Quote(new class String(_tokenString.begin(), _tokenString.end()));
            case NUMBER_TOKEN:
                return new Quote(new class Number(_tokenNumber));
            case '^':
            case '@':
            case '~':
                return getPath(token);
            case '{':
                return getFragment();
            case '!':
                value = getValue();
                if (value) {
                    return new Negative(value);
                } else {
                    error("can not match value for !");
                    return nullptr;
                }
            case '#':
                value = getValue();
                if (value) {
                    return new Quote(value);
                } else {
                    error("can not match value for #");
                    return nullptr;
                }
            case '(':
                return getList();
            default:
                return nullptr;
        }
    }
    
    Value* Source::getKey() {
        auto token = _token;
        nextToken();
        switch (token) {
            case STRING_TOKEN:
                return new Quote(new class String(_tokenString.begin(), _tokenString.end()));
            case '{':
                return getFragment();
            case '(':
                return getList();
            default:
                return nullptr;
        }
    }
    
}